File: src/main/java/com/elissamc/api/MenuSystem/menu/GamesMenu.java
package com.elissamc.api.MenuSystem.menu;

import cn.nukkit.Player;
import cn.nukkit.event.EventHandler;
import cn.nukkit.event.Listener;
import cn.nukkit.event.inventory.InventoryTransactionEvent;
import cn.nukkit.inventory.Inventory;
import cn.nukkit.inventory.InventoryType;
import cn.nukkit.inventory.transaction.action.InventoryAction;
import cn.nukkit.item.Item;
import cn.nukkit.nbt.tag.CompoundTag;
import com.elissamc.ElissaMC;
import com.elissamc.api.MenuSystem.ChestInventory;
import com.elissamc.games.GameParent;

import java.util.Map;

public class GamesMenu implements Listener {

    public static void showMenu(Player player) {
        ChestInventory inv = new ChestInventory(player, InventoryType.CHEST, "games", "Server Menu", 9);
        mainMenu(inv);
        player.addWindow(inv);
    }

    private static void mainMenu(ChestInventory inv) {
        inv.setSize(9);
        int i = 0;
        for (Map.Entry<String, GameParent> game : ElissaMC.plugin.getInstance().getGameHandler().getGames().entrySet()) {
            inv.setItem(i, Item.get(Item.WOOL).setCustomName(game.getKey()).setCustomBlockData(new CompoundTag().putString("game", game.getKey())));
            i++;
        }
    }

    @EventHandler
    public void grabzItemInChest(InventoryTransactionEvent event) {
        for (InventoryAction action : event.getTransaction().getActions()) {
            ChestInventory inventory = null;
            CompoundTag data = action.getSourceItem().getCustomBlockData();
            for (Inventory inv : event.getTransaction().getInventories()) {
                if (inv instanceof ChestInventory) {
                    if (((ChestInventory) inv).getOwner().equals("games")) {
                        event.setCancelled();
                        inventory = (ChestInventory) inv;
                    }
                }
            }

            if (action.getSourceItem().hasCustomBlockData()) {
                assert inventory != null;
                Player player = event.getTransaction().getSource();
                if (data.contains("game")) {
                    String name = data.getString("game");
                    ElissaMC.plugin.getInstance().getGameHandler().joinGame(player, name);
                }
            }
        }
    }
}

File: src/main/kotlin/com/elissamc/api/MenuSystem/menu/GamesMenu.kt
package com.elissamc.api.MenuSystem.menu

import cn.nukkit.Player
import cn.nukkit.event.EventHandler
import cn.nukkit.event.Listener
import cn.nukkit.event.inventory.InventoryTransactionEvent
import cn.nukkit.inventory.Inventory
import cn.nukkit.inventory.InventoryType
import cn.nukkit.inventory.transaction.action.InventoryAction
import cn.nukkit.item.Item
import cn.nukkit.nbt.tag.CompoundTag
import cn.nukkit.utils.DyeColor
import com.elissamc.ElissaMC
import com.elissamc.api.MenuSystem.ChestInventory
import com.elissamc.games.GameParent

class GamesMenu:Listener {

@EventHandler
 fun grabzItemInChest(event:InventoryTransactionEvent) {
for (action in event.transaction.actions)
{
var inventory:ChestInventory? = null
val data = action.sourceItem.customBlockData
for (inv in event.transaction.inventories)
{
if (inv is ChestInventory)
{
if (inv.owner == "games")
{
event.setCancelled()
inventory = inv
}
}
}

if (action.sourceItem.hasCustomBlockData())
{
assert(inventory != null)
val player = event.transaction.source
if (data.contains("game"))
{
val name = data.getString("game")
inventory!!.close(player)
ElissaMC.plugin.instance!!.gameHandler!!.joinGame(player, name)
}
}
}
}

companion object {

 fun showMenu(player:Player) {
val inv = ChestInventory(player, InventoryType.CHEST, "games", "Server Menu", 9)
mainMenu(inv)
player.addWindow(inv)
}

private fun mainMenu(inv:ChestInventory) {
inv.size = 9
var i = 0
for ((key) in ElissaMC.plugin.instance.gameHandler.games)
{
inv.setItem(i, Item.get(Item.WOOL, DyeColor.LIME.woolData).setCustomName(key).setCustomBlockData(CompoundTag().putString("game", key)))
i++
}
}
}
}
